cpu/esp_common/periph/uart.c

```diff
@@ -400,17 +420,27 @@ static int _uart_set_baudrate(uart_t uart, uint32_t baudrate)
 
     _uarts[uart].baudrate = baudrate;
 
+#ifdef MCU_ESP8266
+
+    /* compute and set clock divider */
+    uint32_t clk_div = UART_CLK_FREQ / _uarts[uart].baudrate;
+    _uarts[uart].regs->clk_div.val = clk_div & 0xFFFFF;
+
+#else
+
+/* TODO look for an HAL/LL API function */
```

Comment:
Do you mean the comment? Tt is still a ToDo to change the code to ESP-IDF uart HAL/LL.

Reply:
The code wasn't changed, only the logic was reversed. The former `#else` branch for ESP8266 is now the conditional `#ifdef MCU_ESP8266` and the former `#ifdef MCU_ESP32` is now the `#else` branch good for all ESP32x SoC variants:
```.diff
- #ifdef MCU_ESP32
-     /* use APB_CLK */
-     _uarts[uart].regs->conf0.tick_ref_always_on = 1;
-     /* compute and set the integral and the decimal part */
-     uint32_t clk_div = (UART_CLK_FREQ << 4) / _uarts[uart].baudrate;
-     _uarts[uart].regs->clk_div.div_int  = clk_div >> 4;
-     _uarts[uart].regs->clk_div.div_frag = clk_div & 0xf;
- #else
-     /* compute and set clock divider */
-     uint32_t clk_div = UART_CLK_FREQ / _uarts[uart].baudrate;
-     _uarts[uart].regs->clk_div.val = clk_div & 0xFFFFF;
- #endif
+ #ifdef MCU_ESP8266
+     /* compute and set clock divider */
+     uint32_t clk_div = UART_CLK_FREQ / _uarts[uart].baudrate;
+     _uarts[uart].regs->clk_div.val = clk_div & 0xFFFFF;
+ #else
+     /* TODO look for an HAL/LL API function */
+     #ifdef MCU_ESP32
+         /* use APB_CLK */
+         _uarts[uart].regs->conf0.tick_ref_always_on = 1;
+     #endif
+     #ifdef MCU_ESP32C3
+        _uarts[uart].regs->clk_conf.sclk_sel = 1; /* APB clock used instead of XTAL */
+     #endif
+     /* compute and set the integral and the decimal part */
+     uint32_t clk_div = (UART_CLK_FREQ << 4) / _uarts[uart].baudrate;
+     _uarts[uart].regs->clk_div.div_int  = clk_div >> 4;
+     _uarts[uart].regs->clk_div.div_frag = clk_div & 0xf;
+ #endif

I have indented the nested `#ifdef` for a clearer structure. Within the `#else` branch a difference is made between the ESP32x SoC variants.

Even though ESP32C3 support is not part of this PR, I had used ESP32C3 support to split-off these PRs. However, I was not able to split-off the PRs so that they are completely free of code for ESP32C3. OK, I could remove it from this PR, but I'm afraid I'll forget to add the code for ESP32C3 back later. The PRs are just very complex.

Reply:
Ah so this is not a new TODO but one that was just implicit before?

Reply:
It's just a reminder that the UART implementation doesn't actually use the ESP IDF uart HAL/LL yet, but should in the future.